src/net/client/multi_stream.rs

```diff
@@ -246,9 +301,24 @@ impl<Req: ComposeRequest + Clone + 'static> Request<Req> {
     /// it is resolved, you can call it again to get a new future.
     pub async fn get_response(&mut self) -> Result<Message<Bytes>, Error> {
         loop {
+            let elapsed = self.start.elapsed();
+            if elapsed >= self.conn.response_timeout {
+                return Err(Error::StreamReadTimeout);
+            }
+            let remaining = self.conn.response_timeout - elapsed;
+
             match self.state {
                 QueryState::RequestConn => {
-                    let rx = match self.conn.new_conn(self.conn_id).await {
+                    let to = match timeout(
+                        remaining,
+                        self.conn.new_conn(self.conn_id),
+                    )
+                    .await
+                    {
+                        Err(_) => return Err(Error::StreamReadTimeout),
+                        Ok(v) => v,
+                    };
+                    let rx = match to {
```

Comment:
Use `.map_err()` instead of these large `match` statements -- here and in the rest of this file.